csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+          continue;
+        }
+        merged_pad_width = merged_pad_width == nullptr
+            ? pad_width
+            : add(merged_pad_width, pad_width);
+      }
+      merged_pad_widths.push_back(
+          merged_pad_width == nullptr ? pad_tv->fusion()->zeroVal()
+                                      : merged_pad_width);
+    }
+  }
+
+  // construct TensorDomain for output TV.
+  std::vector<IterDomain*> merged_root_ids;
+  std::vector<IterDomain*> merged_logical_ids;
+  for (const auto i : c10::irange(rank)) {
+    Val* left_pad = merged_pad_widths.at(i * 2);
+    Val* right_pad = merged_pad_widths.at(i * 2 + 1);
+    IterDomain* inp_id = inp_dom.at(i);
+    if (left_pad->isZeroInt() && right_pad->isZeroInt()) {
+      merged_root_ids.push_back(inp_id->cloneWithoutRFactor());
+      merged_logical_ids.push_back(merged_root_ids.back());
+      continue;
+    }
+    // NOTE: nvfuser pad doesn't support negative padding, so we don't have to
+    // worry about it cancelling out.
+    IterDomain* merged_root_id =
+        IterDomainBuilder(inp_id).is_rfactor_domain(true).build();
+    merged_root_ids.push_back(merged_root_id);
+    merged_logical_ids.push_back(IterDomain::resize(
+        merged_root_id,
+        left_pad,
+        right_pad,
+        true,
+        ref_iter_type.at(i)->getIterType()));
+  }
+
+  auto* new_out = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          merged_root_ids, merged_logical_ids, merged_logical_ids),
+      pad_tv->getDataType().value());
+  IrBuilder::create<PadOp>(
+      new_out,
+      pad_tv,
+      merged_pad_widths,
+      SimplifyingIrBuilder::maybeCastExpr(
+          pad_tv->getDataType().value(), pad_value));
+  return new_out;
+}
+
+// Note [ PadOp Propagation Rule ]
+//
+// Push PadOp to its producer to reduce segmentation caused by the resize node
+// introduced in PadOp. The hope is that we can either push PadOp to inputs to
+// the fusion, or far enough that the fusion segments before the PadOp would
+// become a trivial no-op.
+//
+// The concept is that the following program:
+//   tv1 = UnaryOp(tv0)
+//   tv2 = PadOp(tv1)
+// can be replayed as the program below:
+//   tv0_padded = PadOp(tv0)
+//   tv2_new = UnaryOp(tv0_padded)
+// Given certain constraints on the UnaryOp, we can ensure that the
+// computational output remains the same when we replace all uses of `tv2` with
+// `tv2_new`.
+//
+// This function only propagates simple padding, where its pad value is `zero`
+// (or `false` for boolean) and pad widths are non-negative. This allows us to
+// unconditionally merge neighboring PadOps as a single op. We also restrict
+// propagation on operatoins that can allow PadOp to propagated across. See
+// `zeroIsFixedPoint` and `zeroIsIdentity`.
+void propagatePad(Fusion* fusion) {
+  // propagating PadOp
+  auto exprs = fusion->exprs();
+  auto filtered_pads = ir_utils::filterByType<PadOp>(exprs);
+  std::vector<PadOp*> frontier;
```

Comment:
```suggestion
  std::vector<PadOp*> worklist;
```
Nit: move_split_cat processes one `cat` at a time, and I used `frontier` to mean a line of ops that are control-equivalent to the `cat` op being processed. Here, the pads in the list might not form a "frontier" per say.